doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,330 @@
+
+For `C`, `N` and `P`, the entire packet much consist of only frames with the
+marking for the packet to quality for it.  For example, a packet with an ACK
+frame and a _stream_ frame would qualify for neither the `C` or `N` markings.
+
+### Notes
+
+- Will likely need wrapper structure around the individial frames.
+
+### Packets
+
+Frames are coalesced into packets which are then sent by the record layer.
+The `packet_header` is a pointer to the leading bytes of the packet.
+The `frames` are pointers to the individual frames that make up the
+packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+enum packet_validity_e {
+    QUIC_PACKET_INITIAL,
+    QUIC_PACKET_HANDSHAKE,
+    QUIC_PACKET_0_RTT,
+    QUIC_PACKET_1_RTT
+};
+
+typedef enum packet_validity_e PACKET_VALIDITY;
+
+struct ossl_quic_packet_st {
+    SSL *connection;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` struture from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+### Stream Send Buffers
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL_QUIC_TX_PACKETISER *tx, SSL *stream,
+                               const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
+
+#### Notes
+
+- Should more than one frame be queueable at once?
+- More arguments are likely going to be required.
+- How to destroy frames?  `OPENSSL_free` or a callback?
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
```

Comment:
Does the packetiser need to know anything about the inner TLS-1.3 handshake? It has some input "stream" buffers - some of which contain application data, some of which contain crypto data. It just takes the data from the various sources and forms the packet from it. Once the packet is complete it tells the "current" record layer to encrypt and send it.

Reply:
IMO it only needs to know which frames are eligible for being sent as without keys for the appropriate encryption levels you cannot encrypt the packet. So IMO the ordering of the initial, crypto and stream frames is given by the keys available or not.

Reply:
Presumable it (or the ACK Handler) will need to know something about what's going on.  0-RTT frames can morph into 1-RTT frames.  Currently it looks like the TXP will be caching frames for retransmit & I suspect their contents won't be cached anywhere else which means it will have to handle this.